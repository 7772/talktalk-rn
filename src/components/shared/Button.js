// @flow
import React, { Component } from 'react';
import {
  ActivityIndicator,
  StyleSheet,
  TouchableOpacity,
  Image,
  Text,
  View,
} from 'react-native';
import type {
  ____ViewStyleProp_Internal as ViewStyle,
  ____TextStyleProp_Internal as TextStyle,
  ____ImageStyleProp_Internal as ImageStyle,
} from 'react-native/Libraries/StyleSheet/StyleSheetTypes';

import { ratio, colors } from '../../utils/Styles';

type Styles = {
  btn: ViewStyle,
  btnDisabled: ViewStyle,
  txt: TextStyle,
  imgLeft: ImageStyle,
};

const styles: Styles = StyleSheet.create({
  container: {
    flex: 1,
  },
  btn: {
    backgroundColor: 'transparent',
    alignSelf: 'center',
    borderRadius: 4,
    borderWidth: 2,
    width: '100%',
    height: '100%',
    borderColor: 'white',

    alignItems: 'center',
    justifyContent: 'center',
  },
  btnDisabled: {
    backgroundColor: 'rgb(243,243,243)',
    alignSelf: 'center',
    borderRadius: 4,
    borderWidth: 2,
    width: '100%',
    height: '100%',
    borderColor: '#333',

    alignItems: 'center',
    justifyContent: 'center',
  },
  txt: {
    fontSize: 14,
    color: 'white',
  },
  imgLeft: {
    width: 24,
    height: 24,
    position: 'absolute',
    left: 16,
  },
});

type Props = {
  constainerStyle?: ViewStyle;
  isLoading?: boolean;
  isDisabled?: boolean;
  onPress?: () => void;
  style?: ViewStyle;
  disabledStyle?: ViewStyle;
  textStyle?: TextStyle;
  imgLeftSrc?: any;
  imgLeftStyle?: ImageStyle;
  indicatorColor?: string;
  activeOpacity?: number;
  children?: any;
};

type State = {

};

class Button extends Component<Props, State> {
  static defaultProps: Props = {
    isLoading: false,
    isDisabled: false,
    style: styles.btn,
    textStyle: styles.txt,
    imgLeftStyle: styles.imgLeft,
    indicatorColor: 'white',
    activeOpacity: 0.5,
  };

  render() {
    if (this.props.isDisabled) {
      return (
        <View style={this.props.constainerStyle}>
          <View style={this.props.disabledStyle}>
            <Text style={this.props.textStyle}>{this.props.children}</Text>
          </View>
        </View>
      );
    }
    if (this.props.isLoading) {
      return (
        <View style={this.props.constainerStyle}>
          <View style={this.props.style}>
            <ActivityIndicator size='small' color={this.props.indicatorColor} />
          </View>
        </View>
      );
    }
    return (
      <View style={this.props.constainerStyle}>
        <TouchableOpacity
          activeOpacity={this.props.activeOpacity}
          onPress={this.props.onPress}
        >
          <View style={this.props.style}>
            {
              this.props.imgLeftSrc
                ? <Image
                  style={this.props.imgLeftStyle}
                  source={this.props.imgLeftSrc}
                />
                : null
            }
            <Text style={this.props.textStyle}>{this.props.children}</Text>
          </View>
        </TouchableOpacity>
      </View>
    );
  }
}

export default Button;
