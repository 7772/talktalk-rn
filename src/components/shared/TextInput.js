// @flow
import React, { Component } from 'react';
import {
  ActivityIndicator,
  StyleSheet,
  Image,
  Text,
  View,
  TextInput,
} from 'react-native';

import type {
  ____ViewStyleProp_Internal as ViewStyle,
  ____TextStyleProp_Internal as TextStyle,
  ____ImageStyleProp_Internal as ImageStyle,
} from 'react-native/Libraries/StyleSheet/StyleSheetTypes';

import { ratio, colors } from '../../utils/Styles';

type Styles = {
  wrapper: ViewStyle,
  label: ViewStyle,
  labelFocus: TextStyle,
  input: ViewStyle,
  inputFocus: TextStyle,
};

const styles: Styles = StyleSheet.create({
  wrapper: {
    alignSelf: 'stretch',

    flexDirection: 'column',
    justifyContent: 'flex-start',
    alignItems: 'flex-start',
  },
  label: {
    color: colors.blueyGray,
    marginBottom: 8,
    fontSize: 12,
  },
  labelFocus: {
    color: colors.dodgerBlue,
    marginBottom: 8,
    fontSize: 12,
  },
  input: {
    alignSelf: 'stretch',
    color: colors.dusk,
    fontSize: 16,
    paddingVertical: 22,
    paddingHorizontal: 20,
    borderWidth: 1,
    borderColor: 'rgb(233,237,244)',

    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'flex-start',
  },
  inputFocus: {
    alignSelf: 'stretch',
    color: colors.dusk,
    fontSize: 16,
    paddingVertical: 22,
    paddingHorizontal: 20,
    borderWidth: 1,
    borderColor: colors.dodgerBlue,

    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'flex-start',
  },
});

type Props = {
  style?: ViewStyle;
  labelStyle?: TextStyle;
  labelStyleFocus?: TextStyle;
  placeholderTextColor?: string;
  inputStyle?: TextStyle;
  isPassword?: boolean;
  multiline?: boolean;
  txtLabel?: string;
  txtHint?: string;
  txt?: string;
  onFocus?: () => void;
  onTextChanged?: (text: string) => void;
  onSubmitEditing?: (text: string) => void;
  returnKeyType?: any;
};

type State = {
  focused: boolean;
};

class Shared extends Component<Props, State> {
  static defaultProps: Props = {
    style: styles.wrapper,
  };

  state = {
    focused: false,
  };

  constructor(props: Props) {
    super(props);
  }

  render() {
    return (
      <View style={[
        styles.wrapper,
        this.props.style,
      ]}>
        { this.renderTxtLabel() }
        <TextInput
          style={[
            this.state.focused ? styles.inputFocus : styles.input,
            this.props.inputStyle,
          ]}
          underlineColorAndroid='transparent' // android fix
          autoCapitalize='none'
          autoCorrect={false}
          multiline={this.props.multiline}
          onChangeText={this.props.onTextChanged}
          value={this.props.txt}
          onFocus={ () => this.setState({ focused: true }) }
          onBlur={ () => this.setState({ focused: false }) }
          placeholder={this.props.txtHint}
          placeholderTextColor={this.props.placeholderTextColor}
          onSubmitEditing={this.props.onSubmitEditing}
          returnKeyType={this.props.returnKeyType}
          secureTextEntry={this.props.isPassword}
        />
      </View>
    );
  }

  renderTxtLabel = () => {
    if (this.props.txtLabel) {
      return (
        <Text style={this.state.focused ? this.props.labelStyleFocus : this.props.labelStyle}>
          {this.props.txtLabel}
        </Text>
      );
    }
    return null;
  }
}

export default Shared;
