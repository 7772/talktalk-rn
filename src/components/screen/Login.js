// @flow
import React, { Component } from 'react';
import {
  StyleSheet,
  TouchableOpacity,
  Image,
  Text,
  View,
  SafeAreaView,
  ScrollView,
} from 'react-native';

import type {
  ____ViewStyleProp_Internal as ViewStyle,
  ____TextStyleProp_Internal as TextStyle,
  ____ImageStyleProp_Internal as ImageStyle,
} from 'react-native/Libraries/StyleSheet/StyleSheetTypes';

import TextInput from '../shared/TextInput';
import Button from '../shared/Button';
import StatusBar from '../shared/StatusBar';

import { IC_ICON } from '../../utils/Icons';
import { statusBarHeight, ratio, colors } from '../../utils/Styles';
import { getString } from '../../../STRINGS';

type Styles = {
  scrollView: ViewStyle,
  container: ViewStyle,
  iconWrapper: ViewStyle,
  icon: ImageStyle,
  iconTxt: TextStyle,
  wrapper: ViewStyle,
  viewBtnWrapper: ViewStyle,
  btnSignup: ViewStyle,
  txtSignUp: TextStyle,
  btnLogin: ViewStyle,
  txtLogin: TextStyle,
  touchForgotPw: ViewStyle,
  txtForgotPw: TextStyle,
  txtCopyright: TextStyle,
}

export const styles: Styles = StyleSheet.create({
  scrollView: {
    backgroundColor: 'white',
  },
  container: {
    flex: 1,

    flexDirection: 'column',
    alignItems: 'center',
  },
  iconWrapper: {
    position: 'absolute',
    top: 76,
    left: 40,

    flexDirection: 'column',
    alignItems: 'flex-start',
  },
  icon: {
    width: 60,
    height: 60,
  },
  iconTxt: {
    color: colors.dusk,
    fontSize: 20,
    fontWeight: 'bold',
    marginTop: 8,
  },
  wrapper: {
    marginTop: 260,
    width: '78%',
    height: 300,

    flexDirection: 'column',
    alignItems: 'center',
  },
  viewBtnWrapper: {
    alignSelf: 'stretch',
    marginTop: 20,
    height: 60,

    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center',
  },
  btnSignup: {
    backgroundColor: 'transparent',
    borderRadius: 4,
    borderWidth: 1,
    height: '100%',
    width: '100%',
    borderColor: colors.dodgerBlue,

    alignItems: 'center',
    justifyContent: 'center',
  },
  txtSignUp: {
    fontSize: 16,
    fontWeight: 'bold',
    color: colors.dodgerBlue,
  },
  btnLogin: {
    backgroundColor: colors.dodgerBlue,
    borderColor: colors.dodgerBlue,
    borderRadius: 4,
    borderWidth: 1,
    height: 60,
    shadowColor: colors.dodgerBlue,
    shadowOffset: {
      width: 0,
      height: 10,
    },
    shadowRadius: 4,
    shadowOpacity: 0.3,

    alignItems: 'center',
    justifyContent: 'center',
  },
  txtLogin: {
    fontSize: 16,
    fontWeight: 'bold',
    color: 'white',
  },
  touchForgotPw: {
    marginTop: 20,
  },
  txtForgotPw: {
    fontSize: 12,
    color: colors.dodgerBlue,
    textDecorationLine: 'underline',
  },
  txtCopyright: {
    marginTop: 80,
    fontSize: 12,
    color: colors.cloudyBlue,
  },
});

type Props = {
  navigation: any;
};
type State = {
  isLoggingIn: boolean;
  email: string;
  pw: string;
};

class Screen extends Component<Props, State> {
  static navigationOptions = {
    headerStyle: {
      borderBottomWidth: 0,
    }
  };

  constructor(props: Props) {
    super(props);
    this.state = {
      isLoggingIn: false,
      email: '',
      pw: '',
    };
  }

  render() {
    return (
      <ScrollView style={styles.scrollView}>
        <StatusBar isDarkContent={true}/>
        <View style={styles.container}>
          <View style={styles.iconWrapper}>
            <Image style={styles.icon} source={IC_ICON}/>
            <Text style={styles.iconTxt}>{getString('HELLO')}.</Text>
          </View>
          <View style={styles.wrapper}>
            <TextInput
              // txtLabel={ getString('EMAIL') }
              txtHint={ getString('EMAIL') }
              txt={ this.state.email }
              onTextChanged={ (text) => this.onTextChanged('EMAIL', text)}
            />
            <TextInput
              style={{ marginTop: 8 }}
              // txtLabel={ getString('EMAIL') }
              txtHint={ getString('PASSWORD') }
              txt={ this.state.pw }
              onTextChanged={ (text) => this.onTextChanged('PW', text)}
              isPassword={ true }
            />
            <View style={styles.viewBtnWrapper}>
              <Button
                id='signup'
                constainerStyle={{ flex: 1 }}
                onPress={() => this.goToSignup()}
                style={styles.btnSignup}
                textStyle={styles.txtSignUp}
              >{getString('SIGNUP')}</Button>
              <View style={{ width: 8 }}/>
              <Button
                id='login'
                constainerStyle={{ flex: 1 }}
                isLoading={this.state.isLoggingIn}
                onPress={() => this.onLogin()}
                style={styles.btnLogin}
                textStyle={styles.txtLogin}
              >{getString('LOGIN')}</Button>
            </View>
            <TouchableOpacity
              activeOpacity={0.5}
              onPress={() => this.goToFindPw()}
              style={styles.touchForgotPw}
            >
              <Text style={styles.txtForgotPw}>{getString('FORGOT_PW')}</Text>
            </TouchableOpacity>
            <Text style={styles.txtCopyright}>copyright by dooboolab.com</Text>
          </View>
        </View>
      </ScrollView>
    );
  }

  onTextChanged = (type: string, text: string) => {
    switch (type) {
      case 'EMAIL':
        this.setState({ email: text });
        return;
      case 'PW':
        this.setState({ pw: text });
    }
  }

  goToSignup = () => {
    this.props.navigation.navigate('Signup');
    // this.props.navigation.navigate('Signup');
  }
  goToFindPw = () => {
    this.props.navigation.navigate('FindPw');
  }

  onLogin = () => {
    this.setState({ isLoggingIn: true }, () => {
      try {
      } catch (err) {
        // Alert.alert(getString('ERROR'), err.message);
      } finally {
        this.setState({ isLoggingIn: false });
      }
    });
  }
}

export default Screen;
