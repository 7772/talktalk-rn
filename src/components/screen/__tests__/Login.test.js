import 'react-native';
import * as React from 'react';
import Login, { styles } from '../Login';
import  { getString } from '../../../../STRINGS';
import TextInput from '../../shared/TextInput';
import Button from '../../shared/Button';

// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';
import { shallow, render } from 'enzyme';

const wrapper = shallow(<Login />);
const tree = renderer.create(<Login />).toJSON();
const instance = renderer.create(<Login />).getInstance();

describe('rendering test', () => {
  const findElement = function (tree, element) {
    let result = undefined;
    for (let node in tree.children) {
      if (tree.children[node].props.txt==element) {
        result = true;
      }
    }
    return result;
  };

  it('renders as expected', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('renders initial state', () => {
    expect(instance.state.email).toEqual('');
    expect(instance.state.pw).toEqual('');
    expect(instance.state.isLoggingIn).toEqual(false);
  });

  it('find all components', () => {
    expect(wrapper.find('ScrollView')).toHaveLength(1);
    expect(wrapper.find('Text')).toHaveLength(3);
    expect(wrapper.find('StatusBar')).toHaveLength(0);
    expect(wrapper.find(TextInput)).toHaveLength(2);
    expect(wrapper.find(Button)).toHaveLength(2);
    expect(wrapper.find('Text').at(0)).toHaveLength(1);
    expect(wrapper.find('Text').at(0).children).toHaveLength(1);
  });

  it('should have default style', () => {
    expect(wrapper.find('ScrollView').prop('style')).toEqual(styles.scrollView);
    expect(wrapper.find('View').at(0).prop('style')).toEqual(styles.container);
    expect(wrapper.find('View').at(0).childAt(0).prop('style')).toEqual(styles.iconWrapper);
    expect(wrapper.find('View').at(0).childAt(1).prop('style')).toEqual(styles.wrapper);
  });

  it('find elements', () => {
    expect(findElement(tree, getString('EMAIL')));
    expect(findElement(tree, getString('PASSWORD')));
  });

  it('all functions defined', () => {
    expect(instance.onLogin).toBeDefined;
    expect(instance.onTextChanged).toBeDefined;
    expect(instance.goToSignup).toBeDefined;
    expect(instance.goToForgotPw).toBeDefined;
  });

  it('onTextChanged', () => {
    instance.onTextChanged('EMAIL', 'aa@aa.aa');
    expect(instance.state.email).toEqual('aa@aa.aa');
    instance.onTextChanged('PW', 'aaaaaa');
    expect(instance.state.pw).toEqual('aaaaaa');
  });
});

describe('interaction', () => {
  beforeEach(() => {
    wrapper.setProps({
      navigation: {
        navigate: jest.fn(),
      },
    });
  });
  describe('clicking the button', () => {
    let spy1;
    let spy2;
    beforeEach(() => {
      wrapper.instance().goToSignup = jest.fn();
      wrapper.instance().onLogin = jest.fn();
      spy1 = jest.spyOn(wrapper.instance(), 'goToSignup');
      spy2 = jest.spyOn(wrapper.instance(), 'onLogin');
    })
    it('should call onLogin callback', () => {
      const signupBtn = wrapper.find(Button).first();
      const loginBtn = wrapper.find(Button).last();
      expect(signupBtn).toHaveLength(1);
      expect(loginBtn).toHaveLength(1);
      signupBtn.props().onPress();
      console.warn(signupBtn.props().onPress());
      loginBtn.props().onPress();

      // TODO: currently not beening called
      // expect(spy1).toHaveBeenCalled();
      // expect(spy2).toBeCalled();
    });
  });
  afterAll(() => {
    Login.prototype.onLogin.mockRestore();
    Login.prototype.goToSignup.mockRestore();
  });
});