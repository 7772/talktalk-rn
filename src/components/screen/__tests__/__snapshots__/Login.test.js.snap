// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering test renders as expected 1`] = `
<ScrollView
  style={
    Object {
      "backgroundColor": "white",
    }
  }
>
  <Shared
    isDarkContent={true}
  />
  <View
    style={
      Object {
        "alignItems": "center",
        "flex": 1,
        "flexDirection": "column",
        "paddingTop": 20,
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "flex-start",
          "flexDirection": "column",
          "left": 40,
          "position": "absolute",
          "top": 76,
        }
      }
    >
      <Component
        source={
          Object {
            "testUri": "../../../assets/icons/icon.png",
          }
        }
        style={
          Object {
            "height": 48,
            "width": 60,
          }
        }
      />
      <Text
        style={
          Object {
            "color": "rgb(65,77,107)",
            "fontSize": 20,
            "fontWeight": "bold",
            "marginTop": 16,
          }
        }
      >
        Hello
        .
      </Text>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "column",
          "height": 300,
          "marginTop": 260,
          "width": "78%",
        }
      }
    >
      <Shared
        onTextChanged={[Function]}
        style={
          Object {
            "alignItems": "flex-start",
            "alignSelf": "stretch",
            "flexDirection": "column",
            "justifyContent": "flex-start",
          }
        }
        txt=""
        txtHint="Email"
      />
      <Shared
        isPassword={true}
        onTextChanged={[Function]}
        style={
          Object {
            "marginTop": 8,
          }
        }
        txt=""
        txtHint="Password"
      />
      <View
        style={
          Object {
            "alignItems": "center",
            "alignSelf": "stretch",
            "flexDirection": "row",
            "justifyContent": "flex-end",
            "marginTop": 20,
          }
        }
      >
        <Button
          activeOpacity={0.5}
          imgLeftStyle={
            Object {
              "height": 24,
              "left": 16,
              "position": "absolute",
              "width": 24,
            }
          }
          indicatorColor="white"
          isDisabled={false}
          isLoading={false}
          onPress={[Function]}
          style={
            Object {
              "alignItems": "center",
              "alignSelf": "center",
              "backgroundColor": "transparent",
              "borderColor": "rgb(58,139,255)",
              "borderRadius": 4,
              "borderWidth": 1,
              "height": 60,
              "justifyContent": "center",
              "marginRight": 4,
              "width": 136,
            }
          }
          textStyle={
            Object {
              "color": "rgb(58,139,255)",
              "fontSize": 16,
              "fontWeight": "bold",
            }
          }
        >
          SIGN UP
        </Button>
        <Button
          activeOpacity={0.5}
          imgLeftStyle={
            Object {
              "height": 24,
              "left": 16,
              "position": "absolute",
              "width": 24,
            }
          }
          indicatorColor="white"
          isDisabled={false}
          isLoading={false}
          onPress={[Function]}
          style={
            Object {
              "alignItems": "center",
              "alignSelf": "center",
              "backgroundColor": "rgb(58,139,255)",
              "borderColor": "rgb(58,139,255)",
              "borderRadius": 4,
              "borderWidth": 1,
              "height": 60,
              "justifyContent": "center",
              "marginLeft": 4,
              "shadowColor": "rgb(58,139,255)",
              "shadowOffset": Object {
                "height": 10,
                "width": 0,
              },
              "shadowOpacity": 0.3,
              "shadowRadius": 4,
              "width": 136,
            }
          }
          textStyle={
            Object {
              "color": "white",
              "fontSize": 16,
              "fontWeight": "bold",
            }
          }
        >
          LOGIN
        </Button>
      </View>
      <TouchableOpacity
        activeOpacity={0.5}
        onPress={[Function]}
        style={
          Object {
            "marginTop": 20,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "rgb(58,139,255)",
              "fontSize": 12,
              "textDecorationLine": "underline",
            }
          }
        >
          Forgot password?
        </Text>
      </TouchableOpacity>
      <Text
        style={
          Object {
            "color": "rgb(175,194,219)",
            "fontSize": 12,
            "marginTop": 80,
          }
        }
      >
        copyright by dooboolab.com
      </Text>
    </View>
  </View>
</ScrollView>
`;
